Terminal.Gui/View/ViewScrollBar.cs

```diff
@@ -0,0 +1,445 @@
+            {
+                return;
+            }
+
+            view._scrollBarType = value;
+            view.DisposeScrollBar ();
+
+            switch (view._scrollBarType)
+            {
+                case ScrollBarType.Vertical:
+                    view._scrollBar = new ScrollBarView { Orientation = Orientation.Vertical };
+
+                    break;
+                case ScrollBarType.Horizontal:
+                    view._scrollBar = new ScrollBarView { Orientation = Orientation.Horizontal };
+
+                    break;
+                case ScrollBarType.Both:
+                    view._scrollBar = new ScrollBarView { Orientation = Orientation.Vertical };
+                    view._scrollBar.OtherScrollBarView = new ScrollBarView { Orientation = Orientation.Horizontal, OtherScrollBarView = view._scrollBar };
+
+                    break;
+                case ScrollBarType.None:
+                    return;
+                default:
+                    throw new ArgumentOutOfRangeException ();
+            }
+
+            Add (view._scrollBar);
+            view.AddEventHandlersForScrollBars (view._scrollBar);
+            view.AddKeyBindingsForScrolling (view._scrollBar);
+
+            if (view._scrollBar.OtherScrollBarView != null)
+            {
+                view.AddKeyBindingsForScrolling (view._scrollBar.OtherScrollBarView);
+            }
+
+            view.SetNeedsDisplay ();
+        }
+    }
+
+    /// <summary>Get or sets if the view-port is kept always visible in the area of this <see cref="ScrollBarView"/></summary>
+    public bool ScrollKeepContentAlwaysInViewPort
+    {
+        get => _scrollBar.KeepContentAlwaysInViewPort;
+        set => _scrollBar.KeepContentAlwaysInViewPort = value;
+    }
+
+    /// <summary>Represent a vertical or horizontal ScrollBarView other than this.</summary>
+    public ScrollBarView ScrollOtherScrollBarView
+    {
+        get => _scrollBar.OtherScrollBarView;
+        set => _scrollBar.OtherScrollBarView = value;
+    }
+
+    /// <summary>Gets or sets the visibility for the vertical or horizontal scroll indicator.</summary>
+    /// <value><c>true</c> if show vertical or horizontal scroll indicator; otherwise, <c>false</c>.</value>
+    public bool ScrollShowScrollIndicator
+    {
+        get => _scrollBar.ShowScrollIndicator;
+        set => _scrollBar.ShowScrollIndicator = value;
+    }
+
+    private void AddEventHandlersForScrollBars (ScrollBarView scrollBar)
+    {
+        if (scrollBar is null)
+        {
+            return;
+        }
+
+        scrollBar.ChangedPosition += ScrollBar_ChangedPosition;
+
+        if (_scrollBar.OtherScrollBarView != null)
+        {
+            _scrollBar.OtherScrollBarView.ChangedPosition += OtherScrollBarView_ChangedPosition;
+        }
+    }
+
+    private void AddKeyBindingsForScrolling (ScrollBarView scrollBar)
+    {
+        if (scrollBar.Orientation == Orientation.Vertical)
```

Comment:
I have two issues with this code.

1) I am not sure there's really a need to have these key bindings added to the `ScrollBar`s at all. I don't think the `ScrollBar`s of `View` should ever actually get focus. they should DRAW like they have focus, when the `View` is focused, but should otherwise ignore keyboard input, letting the `View` handled it. 

2) You test `Orientation` repeatedly. This code (if needed at all) could be much simpler.

Reply:
The `ScrollBar` doesn't really get the focus because is just the `MouseGrabView` working.

Yes I can simplify that repeatedly check- I'll change it.

Reply:
> What other property in View could be confused by this?

That code will be changed. Not finished yet.

Reply:
> Fine. Then check for is not Adornment instead.

R# complains by using `is not` and prefer the `is { }`.
I think only check if `Margin is Adornment` doesn't do much sense to me because Margin is a `Adornment` type already. But I could be wrong and both sentences may achieve the same result.

Reply:
> If the view want to handle the `ScrollBar`'s key bindings, it just only have to add their ones and they will be replaced.

If these keybindings are for when `ScrollBar` I used outside of `View` (standalone), then why are they set in `View`? Why are they not set in `ScrollBar.cs`?

Reply:
>> Fine. Then check for is `not Adornment` instead.

> I think only check if Margin is Adornment doesn't do much sense to me because Margin is a Adornment type already. But I could be wrong and both sentences may achieve the same result.

They do achieve the same thing. However one introduces an unnecessary and confusing coupling to `Margin`. One is clear what the test is for.

Reply:
> > If these keybindings are for when ScrollBar I used outside of View (standalone), then why are they set in View? Why are they not set in ScrollBar.cs?
> 
> First we need to define what is `ScrollBar.cs`. Now I only have `ViewScrollBar.cs`, `ScrollBarView.cs` and `ScrollView.cs`. What is `ScrollBar.cs` on those files? These key bindings is only for to be used when a `View` doesn't define any key binding and want to use scroll bars ( `scrollBar.KeyBindings.Add`). If a View doesn't use scroll bars those key bindings aren't assigned.

When we're done, there should be 

- a new class named `ScrollBar` in `ScrollBar.cs`.  It is a stand alone view that draws either a horiztonal or vertical scrollbar and handles mouse/keyboard input. Just like `Slider`. 
- a new file named `ViewScrollBar` that contains the `View` methods/properties that expose on `View` (as `partial class View`).
- a old, legacy class, named `ScrollBarView` in `ScrollBarView.cs`. It will eventually be deleted. 
- a old, legacy class, named `ScrollView` in `ScrollView.cs`. It will be eventually deleted.
-

Reply:
`View.ScrolllBar` is the same as using the path for namespace which is you really want to avoid, for not have breaking changes if a file is moved to another path.

Reply:
Wouldn't it be better to create a `ScrollBar` folder in ./Views, creating `ScrollBar.cs` there and also moving `ViewScrollBar.cs` there?